org-code-javabuilder/protocol/src/main/java/org/code/protocol/JavabuilderThrowableMessageUtils.java

```diff
@@ -14,15 +15,32 @@ public static JavabuilderThrowableMessage getExceptionMessage(
       Throwable throwable, Enum key, String fallbackMessage) {
     HashMap<String, String> detail = new HashMap<>();
     detail.put(ClientMessageDetailKeys.CONNECTION_ID, Properties.getConnectionId());
-    if (throwable.getCause() != null) {
-      detail.put(ClientMessageDetailKeys.CAUSE, getLoggingString(throwable));
-      if (throwable.getCause().getMessage() != null) {
-        detail.put(ClientMessageDetailKeys.CAUSE_MESSAGE, throwable.getCause().getMessage());
+
+    Throwable cause = throwable.getCause();
+    if (cause != null && cause instanceof InvocationTargetException) {
+      cause = cause.getCause();
+    }
+
+    String preferredFallbackMessage = null;
+    if (cause != null) {
+      // To do: remove CAUSE key once we're using STACK_TRACE and EXCEPTION_TYPE
+      // in (partially) translated exception messages in Javalab.
+      // https://codedotorg.atlassian.net/browse/JAVA-439
+      preferredFallbackMessage = getDefaultFallbackMessageString(cause);
+      detail.put(ClientMessageDetailKeys.CAUSE, preferredFallbackMessage);
+      detail.put(ClientMessageDetailKeys.STACK_TRACE, getUserFacingStackTraceString(cause));
+      detail.put(ClientMessageDetailKeys.EXCEPTION_MESSAGE, getExceptionMessageString(cause));
+      if (cause.getMessage() != null) {
+        detail.put(ClientMessageDetailKeys.CAUSE_MESSAGE, cause.getMessage());
       }
     }
 
+    // Caller providing a fallback message overrides the default generated in this method.
     if (fallbackMessage != null) {
-      detail.put(ClientMessageDetailKeys.FALLBACK_MESSAGE, fallbackMessage);
```

Comment:
nit: should this check happen earlier so that we don't generate a preferredFallbackMessage if we don't need to?